sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -144,6 +145,16 @@ object TypeCoercion {
       .orElse((t1, t2) match {
         case (ArrayType(et1, containsNull1), ArrayType(et2, containsNull2)) =>
           findWiderTypeForTwo(et1, et2).map(ArrayType(_, containsNull1 || containsNull2))
+        case (st1 @ StructType(fields1), st2 @ StructType(fields2)) if st1.sameType(st2) =>
```

Comment:
I don't think we should put this logic here, we are not finding a "wider" type here. I think a more proper place is in `findTightestCommonType` when we deal with data type equality. More generally, I think we should revisit when we should use `==` and when use `DataType.sameType` for data type equality check. For example, we can eliminate struct type cast if unnecessary.

In this PR, even you fixed the struct type equality in type coercion, the `Cast` doesn't know it and we will end up executing unnecessary casts.